addons/sms/models/mail_thread.py

```diff
@@ -24,36 +24,201 @@ def _get_default_sms_recipients(self):
+        :param subtype_id: mail.message.subtype used in mail.message associated
+          to the sms notification process;
+        :param partner_ids: if set is a record set of partners to notify;
+        :param number_field: if set is a name of field to use on current record
+          to compute a number to notify;
+        :param numbers: if set is a list of phone numbers to notify;
+        """
+        self.ensure_one()
+
+        if number_field or (partner_ids is False and numbers is False):
+            info = self._sms_get_recipients_info(force_field=number_field)[self.id]
+            info_partner_ids = info['partner'].ids if info['partner'] else False
+            info_number = info['sanitized'] if info['sanitized'] else info['number']
+            if info_partner_ids and info_number:
+                pid_to_number = self.env.context.get('notify_sms_pid_to_number', {})
+                pid_to_number[info_partner_ids[0]] = info_number
+                self = self.with_context(notify_sms_pid_to_number=pid_to_number)
+            if info_partner_ids:
+                partner_ids = info_partner_ids + (partner_ids or [])
+            if info_number and not info_partner_ids:
+                numbers = [info_number] + (numbers or [])
+
+        if numbers:
+            self = self.with_context(notify_sms_numbers=numbers)
+
+        if subtype_id is False:
+            subtype_id = self.env['ir.model.data'].xmlid_to_res_id('mail.mt_comment')
+
+        return self.message_post(
+            body=body, partner_ids=partner_ids or [],  # TDE FIXME: temp fix otherwise crash mail_thread.py
+            message_type='sms', subtype_id=subtype_id,
+        )
+
+    @api.multi
+    def _notify_thread(self, message, msg_vals=False, model_description=False, mail_auto_delete=True):
+        recipients_data = super(MailThread, self)._notify_thread(
+            message, msg_vals=msg_vals, model_description=model_description,
+            mail_auto_delete=mail_auto_delete,
+        )
+        self._notify_records_by_sms(message, recipients_data, msg_vals=msg_vals, put_in_queue=False)
+        return recipients_data
+
+    @api.multi
+    def _notify_records_by_sms(self, message, recipients_data, msg_vals=False, put_in_queue=False):
+        """ Notification method: by SMS. Some specific context keys allow to customize
+        the process:
+
+          * notify_check_existing: check existing notifications to update instead
+              of creating new one (allow to speedup classic post by always creating
+              new one);
+          * notify_sms_pid_to_number: force a number to notify for a given partner ID
+              instead of taking its mobile / phone number;
+          * notify_sms_numbers: additional numbers to notify;
         """
-        if not numbers:
-            if not partners:
-                partners = self._get_default_sms_recipients()
+        notify_check_existing = self.env.context.get('notify_check_existing', False)
```

Comment:
As this a new dedicated method, it is strange to use context keys: isn't it possible to make them proper arguments. Maybe this means that `_notify_thread` should allow `kwargs` like `message_post`.

Reply:
Well at first try I added explicit parameters, then kwargs, then fall back on context. I am still unsure what is better.

Adding explicit parameters adds noise to notify_thread, as you may have parameters for each notification method. Adding kwargs is actually a bit like context, you still have to fetch the data in a dictionary, either kwargs or the context. Kwargs is a bit more verbose when doing the call so maybe it is better (self._notify(..., my_param=orval) instead of self.with_context(my_param=orval)._notify()).

Additional parameters used in notify by sms are a bit side-parameters: not the primary one. Some side cases: additional numbers to notify, telling we are in "re-send" mode, ...

Bref, not sure. What would make you happy ?